src/oauth2c_token_cache.erl

```diff
@@ -0,0 +1,155 @@
+                   #{default_ttl => DefaultTTL}, []).
+
+-spec start_link() -> {atom(), pid()}.
+start_link() ->
+  start_link(?DEFAULT_TTL).
+-spec start_link(non_neg_integer()) -> {atom(), pid()}.
+start_link(DefaultTTL) ->
+  gen_server:start_link({local, ?SERVER}, ?SERVER,
+                        #{default_ttl => DefaultTTL}, []).
+
+-spec get(integer()) -> {error, atom()} | {ok, term()}.
+get(Key) ->
+  get_token(Key).
+
+-spec set_and_get(Key, LazyValue) -> Value | Error when
+    Key :: integer(),
+    LazyValue :: fun(() -> Value | Error),
+    Value :: {ok, term()},
+    Error :: {error, binary()}.
+set_and_get(Key, LazyValue) ->
+  set_and_get(Key, LazyValue, []).
+
+-spec set_and_get(Key, LazyValue, Options) -> Value | Error when
+    Key :: integer(),
+    LazyValue :: fun(() -> Value | Error),
+    Options :: list(),
+    Value :: {ok,  term()},
+    Error :: {error, atom()}.
+set_and_get(Key, LazyValue, Options) ->
+  gen_server:call(?SERVER, {set_and_get, Key, LazyValue, Options}).
+
+-spec clear() -> true.
+clear() ->
+  ets:delete_all_objects(?TOKEN_CACHE_ID).
+
+%%%_ * gen_server callbacks --------------------------------------------
+
+init(State) ->
+  EtsOpts = [set, public, named_table, {read_concurrency, true}],
+  ets:new(?TOKEN_CACHE_ID, EtsOpts),
+  {ok, State}.
+
+handle_call({set_and_get, Key, LazyValue, Options}, _From,
+  State = #{default_ttl := DefaultTTL}) ->
+  % ForceUpdateLimit is used to solve a race-condition
+  % that occurs when multiple processes are trying to
+  % replace an old token (i.e. the new token has a larger
+  % expiry time than the old token).
+  ForceUpdateLimit =
+    proplists:get_value(force_update_entries_older_or_equal_than,
+                        Options,
+                        undefined),
+  case get_token(Key, ForceUpdateLimit) of
+    {ok, Result} ->
+      {reply, {ok, Result}, State};
+    {error, not_found} ->
+      case LazyValue() of
+        {ok, Result, ExpiryTime0} ->
+          ExpiryTime = get_expiry_time(ExpiryTime0, DefaultTTL),
+          ets:insert(?TOKEN_CACHE_ID,
+            {Key, Result, ExpiryTime}),
+          {reply, {ok, Result}, State};
+        {error, Reason} -> {reply, {error, Reason}, State}
+      end
+  end.
+
+handle_cast(_, State) -> {noreply, State}.
+
+%%%_ * Private functions -----------------------------------------------
+
+get_token(Key) ->
+  get_token(Key, undefined).
+get_token(Key, ExpiryTimeLowerLimit) ->
+  Now = erlang:system_time(second),
+  case ets:lookup(?TOKEN_CACHE_ID, Key)
+  of
+    % Only return cache entry if
+    % (1) It has not expired
+    % (2) Its expiry time is greater than ExpiryTimeLowerLimit
+    [{Key, Result, ExpiryTime}] when ExpiryTime > Now
+                                     andalso
+                                     (ExpiryTimeLowerLimit =:= undefined
+                                     orelse
+                                     ExpiryTime > ExpiryTimeLowerLimit) ->
```

Comment:
Incorrect indentation. This 2 lines need to be aligned with `ExpiryTimeLowerLimit`.

Reply:
I push a commit to fix it